Transfer scd and scd2 for YDB (#23454)
ydb-platform/ydb: ydb/docs/ru/core/dataeng/transfer/scd2.md

        "change_time":"2025-08-22T17:21:18.357503Z",
        "id":"CUSTOMER_1001",
        "attribute2":"Los Angeles 2"
      },
      "before":{
        "attribute1":"John Doe 2",
        "change_time":"2025-08-22T17:12:51.502806Z",
        "id":"CUSTOMER_1001",
        "attribute2":"Los Angeles 2"
      }
    }
  }
  ```

  **3. Удаление записи (создаст событие с `op = "d"`):**

  ```sql
  -- Удаление записи
  DELETE FROM source_customers2 WHERE id = 'CUSTOMER_1002';
  ```

  Это действие создаст CDC-событие примерно следующего вида:

  ```json
  {
    "payload":{
      "op":"d",
      "source":{
        "txId":18446744073709551615,
        "connector":"ydb",
        "version":"1.0.0",
        "step":1755883827000,
        "ts_ms":1755883827948,
        "snapshot":false
      },
      "before":{
        "attribute1":"John Doe",
        "change_time":"2025-08-22T17:28:03.648313Z",
        "id":"CUSTOMER_1002",
        "attribute2":"Los Angeles"
      }
    }
  }
  ```

  Эти CDC-события будут записаны в топик YDB, откуда их будет читать трансфер и преобразовывать в записи таблицы `dimension_scd2` с помощью lambda-функции, описанной выше. Обратите внимание, что:

>   * При создании записи (`op = "c"`) данные берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`
>   * При обновлении записи (`op = "u"`) данные также берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`
>   * При удалении записи (`op = "d"`) данные берутся из поля `before`, а в поле `deleted_at` устанавливается системный момент времени получения информации про удаление записи

Reviewer: ```suggestion
  * При создании записи (`op = "c"`) данные берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`.
  * При обновлении записи (`op = "u"`) данные также берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`.
  * При удалении записи (`op = "d"`) данные берутся из поля `before`, а в поле `deleted_at` устанавливается системный момент времени получения информации про удаление записи.
```

>   ## Пример запроса для получения актуальных данных

Reviewer: ```suggestion
  ## Пример запроса для получения актуальных данных
  
```

>   Для получения данных из SCD2 append only структур хранения можно использовать следующий запрос, который получает данные на момент времени `2025-08-22 19:11:30`:

Reviewer: ```suggestion
  Для получения данных из структур хранения SCD2 append only можно использовать следующий запрос, который получает данные на момент времени `2025-08-22 19:11:30`:
```

Author: нет, так нельзя переписать. я сложно написал, но имел в виду, что "если вы хотите почитать из scd2 данные, то нельзя просто написать select *, а нужно более сложный вопрос"

Reviewer 2: давайте хотя б «структуру хранения» на что-то по-проще вроде «таблица» заменим

  ```sql
  DECLARE $as_of AS Timestamp;
  $as_of = Datetime::MakeDatetime(DateTime::Parse("%Y-%m-%d %H:%M:%S")("2025-08-22 19:11:30"));

  SELECT
    id,
    attribute1,
    attribute2,
    valid_from,
    deleted_at
  FROM (
    SELECT
      t.*,
      ROW_NUMBER() OVER (PARTITION BY id ORDER BY valid_from DESC) AS rn
    FROM dimension_scd2 AS t
    WHERE
      t.valid_from <= $as_of
      AND (t.deleted_at IS NULL OR t.deleted_at > $as_of)  -- запись ещё не удалена к моменту X
  )
  WHERE rn = 1;
  ```

- SCD2 с помощью таблицы хранения промежуточных данных и операции слияния

>   Для работы с CDC данными YDB в формате Debezium при реализации SCD2:

Reviewer: ```suggestion
  Для работы с CDC данными {{ ydb-short-name }} в формате Debezium при реализации SCD2:
```

    $is_deleted = IF($operation == "d", TRUE, FALSE);

    $operation_type = case $operation
                            WHEN 'd' THEN "DELETE"
                            WHEN 'u' THEN "UPDATE"
                            WHEN 'c' THEN "CREATE"
                            else NULL
                        END;

    -- Получаем данные в зависимости от типа операции
    $data = IF($is_deleted, $cdc_data.payload.before, $cdc_data.payload.after);

    $result =  IF($data is not NULL, <|
            id: Unwrap(CAST(Yson::ConvertToString($data.id) AS Utf8)),
            attribute1: CAST(Yson::ConvertToString($data.attribute1) AS Utf8),
            attribute2: CAST(Yson::ConvertToString($data.attribute2) AS Utf8),
            change_time : Unwrap(cast(datetime::MakeDatetime(datetime::ParseIso8601(Yson::ConvertToString($data.change_time))) as timestamp)),
            operation: cast($operation_type as Utf8)
        |>, NULL);

    return $result;
  };

  CREATE TRANSFER dimension_scd2_cdc_changes
        FROM `source_customers2/updates` TO dimension_scd_changes USING $transformation_lambda
        with(FLUSH_INTERVAL=Interval("PT1S"))
        ;

  ```

>   4. Создайте периодический запроса подмерживания изменений из таблицы `dimension_scd_changes` в основную таблицу хранения данных в формате SCD2.

Reviewer: ```suggestion
  4. Создайте периодический запрос для подмерживания изменений из таблицы `dimension_scd_changes` в основную таблицу хранения данных в формате SCD2.
```

# Реализация SCD2 (append-only) с использованием трансфера

## Особенности SCD2 и подход append-only

SCD2 (Type 2) — это подход, при котором при изменении атрибута измерения создаётся новая запись, а старая помечается как неактуальная. Таким образом, сохраняется история изменений. Этот подход используется, когда:
- требуется отслеживать историю изменений данных;
- необходимо выполнять анализ данных с учётом временных периодов;
- важно сохранять аудиторский след изменений;

> - требуется возможность восстановления состояния данных на определённый момент времени.

Reviewer 2: порекомендовать SCD1 со ссылкой, если история не нужна

## Реализация SCD2 с помощью трансфера {#scd2}

  {% note warning %}

  Transfer не может удалять строки из таблиц, а может выполнять только UPSERT в таблицу-приёмник.

>   Поэтому с помощью Transfer можно реализовать только SCD2 append only: удаление помечается флагом и сохраняется момент удаления. Если нужна физическая очистка данных, ее можно выполнять или через явную команду DELETE или с помощью TTL.

Reviewer 2: и тут тоже см. комментарии к scd1

  {% endnote %}

{% list tabs %}

- SCD2 append only

  1. Создайте таблицу:

  ```sql
    CREATE TABLE dimension_scd2 (
        id Utf8 NOT NULL,
        attribute1 Utf8,
        attribute2 Utf8,
        valid_from Timestamp NOT NULL,
        deleted_at Timestamp ,
        PRIMARY KEY (valid_from, id)
    )
    PARTITION BY HASH(valid_from, id)
    WITH(STORE=COLUMN)
  ```

>   2. Создайте lambda-функцию для обработки CDC-данных и трансфер:

Reviewer 2: тут тоже многие комментарии из scd1 применимы

  ```sql
  $transformation_lambda = ($msg) -> {
        $cdc_data = CAST($msg._data AS JSON);

        -- Определяем тип операции
        $operation = Yson::ConvertToString($cdc_data.payload.op);
        $is_deleted = IF($operation == "d", TRUE, FALSE);

        -- Получаем данные в зависимости от типа операции
        $data = IF($is_deleted, $cdc_data.payload.before, $cdc_data.payload.after);

        $result =  IF($data is not NULL, <|
                id: Unwrap(CAST(Yson::ConvertToString($data.id) AS Utf8)),
                attribute1: CAST(Yson::ConvertToString($data.attribute1) AS Utf8),
                attribute2: CAST(Yson::ConvertToString($data.attribute2) AS Utf8),

>                 deleted_at: IF($is_deleted, DateTime::FromMilliseconds(Yson::ConvertToUint64($cdc_data.payload.source.ts_ms)), NULL),

Reviewer 2: ```suggestion
                deleted_at: IF($is_deleted, CAST(Json::ConvertToUint64($cdc_data.payload.source.ts_ms) AS Datetime)),
```

                valid_from : Unwrap(cast(datetime::MakeDatetime(datetime::ParseIso8601(Yson::ConvertToString($data.change_time))) as timestamp)),
            |>, NULL);

>         return $result;

Author: спасибо за замечание. у меня были варианты, когда нужно было возвращать несколько записей сразу: закрывающую и новую, но потом я понял, что не помогает. пример переделать забыл

        "snapshot":false
      },
      "before":{
        "attribute1":"John Doe",
        "change_time":"2025-08-22T17:28:03.648313Z",
        "id":"CUSTOMER_1002",
        "attribute2":"Los Angeles"
      }
    }
  }
  ```

  Эти CDC-события будут записаны в топик YDB, откуда их будет читать трансфер и преобразовывать в записи таблицы `dimension_scd2` с помощью lambda-функции, описанной выше. Обратите внимание, что:

  * При создании записи (`op = "c"`) данные берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`.
  * При обновлении записи (`op = "u"`) данные также берутся из поля `after`, в поле `valid_from` сохраняется значение момента времени изменения записи из поля `changed_at`.
  * При удалении записи (`op = "d"`) данные берутся из поля `before`, а в поле `deleted_at` устанавливается системный момент времени получения информации про удаление записи.

  ## Пример запроса для получения актуальных данных
  

  Для получения данных из SCD2 append only структур хранения можно использовать следующий запрос, который получает данные на момент времени `2025-08-22 19:11:30`:

  ```sql
  DECLARE $as_of AS Timestamp;

>   $as_of = Datetime::MakeDatetime(DateTime::Parse("%Y-%m-%d %H:%M:%S")("2025-08-22 19:11:30"));

Reviewer 2: ты же история, что и в scd1

  SELECT
    id,
    attribute1,
    attribute2,
    valid_from,
    deleted_at
  FROM (
    SELECT
      t.*,

>       ROW_NUMBER() OVER (PARTITION BY id ORDER BY valid_from DESC) AS rn

Author: такого фокуса не знал, спасибо

    FROM dimension_scd2 AS t
    WHERE
      t.valid_from <= $as_of
      AND (t.deleted_at IS NULL OR t.deleted_at > $as_of)  -- запись ещё не удалена к моменту X
  )
  WHERE rn = 1;
  ```

> - SCD2 с помощью таблицы хранения промежуточных данных и операции слияния

Reviewer 2: как-будто плюс-минус полный дубль той второй статьи, которая `../scd2.md` — зачем он тут?

Author: это не дубль. тут фокус на трансфер, а структуры scd2 вторичны

а в ../scd2.md фокус на scd2 и там не описывается трансфер

я сначала писал в формате общей статьи, но потом понял, что трансфер ценен сам по себе, а scd2 ценен тоже сам по себе, поэтому поделил на 2 части

        PRIMARY KEY (change_time, id)
    )
    PARTITION BY HASH(change_time, id)
    WITH(STORE=COLUMN)
  ```

  1. Создайте таблицу для хранения данных:

  ```sql
    CREATE TABLE dimension_scd2_final (
        id Utf8 NOT NULL,
        attribute1 Utf8,
        attribute2 Utf8,
        valid_from Timestamp NOT NULL,
        valid_to Timestamp,
        is_current Timestamp ,
        is_deleted Uint8,
        PRIMARY KEY (valid_from, id)
    )
    PARTITION BY HASH(valid_from, id)
    WITH(STORE=COLUMN)
  ```

  1. Создайте трансфер для переноса данных:

  ```sql
  $transformation_lambda = ($msg) -> {
    $cdc_data = CAST($msg._data AS JSON);

    -- Определяем тип операции
    $operation = Yson::ConvertToString($cdc_data.payload.op);
    $is_deleted = IF($operation == "d", TRUE, FALSE);

    $operation_type = case $operation
                            WHEN 'd' THEN "DELETE"
                            WHEN 'u' THEN "UPDATE"
                            WHEN 'c' THEN "CREATE"
                            else NULL
                        END;

    -- Получаем данные в зависимости от типа операции
    $data = IF($is_deleted, $cdc_data.payload.before, $cdc_data.payload.after);

>     $result =  IF($data is not NULL, <|

Author: так записи пропадут. лучше будет ошибка

            id: Unwrap(CAST(Yson::ConvertToString($data.id) AS Utf8)),
            attribute1: CAST(Yson::ConvertToString($data.attribute1) AS Utf8),
            attribute2: CAST(Yson::ConvertToString($data.attribute2) AS Utf8),
            change_time : Unwrap(cast(datetime::MakeDatetime(datetime::ParseIso8601(Yson::ConvertToString($data.change_time))) as timestamp)),

>             operation: cast($operation_type as Utf8)

Reviewer 2: ```suggestion
            operation: $operation_type
```
